data_model/derive/src/event_set.rs

```diff
@@ -0,0 +1,395 @@
+                }
+            },
+        );
+
+        let doc = format!(" An event set for [`{event_enum_ident}`]s\n\nEvent sets of the same type can be combined with a custom `|` operator");
+
+        tokens.extend(quote! {
+            #[derive(
+                Copy,
+                Clone,
+                PartialEq,
+                Eq,
+                PartialOrd,
+                Ord,
+                Hash,
+                // this introduces tight coupling with these crates
+                // but it's the easiest way to make sure those traits are implemented
+                parity_scale_codec::Decode,
+                parity_scale_codec::Encode,
+                // TODO: we probably want to represent the bit values for each variant in the schema
+                iroha_schema::IntoSchema,
+            )]
+            #[repr(transparent)]
+            #[doc = #doc]
+            #vis struct #set_ident(u32);
+
+            // we want to imitate an enum here, so not using the SCREAMING_SNAKE_CASE here
+            #[allow(non_upper_case_globals)]
+            impl #set_ident {
+                #( #flag_defs )*
+            }
+
+            impl #set_ident {
+                /// Creates an empty event set
+                pub const fn empty() -> Self {
+                    Self(0)
+                }
+
+                /// Creates an event set containing all events
+                pub const fn all() -> Self {
+                    Self(
+                        #(
+                            #flag_idents.0
+                        )|*
+                    )
+                }
+
+                /// Combines two event sets by computing a union of two sets
+                ///
+                /// A const method version of the `|` operator
+                pub const fn or(self, other: Self) -> Self {
+                    Self(self.0 | other.0)
+                }
+
+                /// Checks whether an event set is a superset of another event set
+                ///
+                /// That is, whether `self` will match all events that `other` contains
+                pub const fn contains(&self, other: Self) -> bool {
+                    (self.0 & other.0) == other.0
+                }
+
+                /// Decomposes an `EventSet` into a vector of basis `EventSet`s, each containing a single event
+                ///
+                /// Each of the event set in the vector will be equal to some of the associated constants for the `EventSet`
+                pub fn decompose(&self) -> Vec<Self> {
+                    let mut result = Vec::new();
+
+                    #(if self.contains(#flag_idents) {
+                        result.push(#flag_idents);
+                    })*
+
+                    result
+                }
+
+                /// Checks whether an event set contains a specific event
+                pub const fn matches(&self, event: &#event_enum_ident) -> bool {
```

Comment:
```suggestion
                #[cfg(feature = "transparent_api")]
                pub const fn matches(&self, event: &#event_enum_ident) -> bool {
```

Reply:
While this does feel like something that should be gated behind `transparent_api` feature, but this wouldn't allow us to test it that easily (https://github.com/hyperledger/iroha/blob/eaf54e724083a397db8477254e54866810914b40/data_model/derive/tests/event_set.rs). We would either have to introduce a `transparent_api` on the `data_model_derive` crate, create a separate crate to test this macro or introduce additional logic to conditionally emit the `transparent_api` gate depending on derive macro params.

Do you find it that important that this method should be gated?

Reply:
I found it weird that you've put these tests in the `data_model/derive` and not int `data_model` crate. Was there any particular reason for that?

Reply:
The reason is that I wanted to test the derive macro itself, isolated from the rest of `data_model`. And, conceptually, it tests the derive macro, not anything we _actually_ expose from `data_model`.

Making sure the macro can be used in a "clean" environment also allows for ui tests (though I don't think they are necessary in this particular case)